server/sonar-db-dao/src/main/java/org/sonar/db/measure/ProjectMeasuresIndexerIterator.java

```diff
@@ -0,0 +1,331 @@
+  private final Iterator<Project> projects;
+
+  private ProjectMeasuresIndexerIterator(PreparedStatement measuresStatement, Map<Long, String> metricKeysByIds, List<Project> projects) throws SQLException {
+    this.measuresStatement = measuresStatement;
+    this.metricKeysByIds = metricKeysByIds;
+    this.projects = projects.iterator();
+  }
+
+  public static ProjectMeasuresIndexerIterator create(DbSession session, long afterDate, @Nullable String projectUuid) {
+    try {
+      Map<Long, String> metrics = selectMetricKeysByIds(session);
+      List<Project> projects = selectProjects(session, afterDate, projectUuid);
+      PreparedStatement projectsStatement = createMeasuresStatement(session, metrics.keySet());
+      return new ProjectMeasuresIndexerIterator(projectsStatement, metrics, projects);
+    } catch (SQLException e) {
+      throw new IllegalStateException("Fail to execute request to select all project measures", e);
+    }
+  }
+
+  private static Map<Long, String> selectMetricKeysByIds(DbSession session) {
+    Map<Long, String> metrics = new HashMap<>();
+    try (PreparedStatement stmt = createMetricsStatement(session);
+      ResultSet rs = stmt.executeQuery()) {
+      while (rs.next()) {
+        metrics.put(rs.getLong(1), rs.getString(2));
+      }
+      return metrics;
+    } catch (SQLException e) {
+      throw new IllegalStateException("Fail to execute request to select all metrics", e);
+    }
+  }
+
+  private static PreparedStatement createMetricsStatement(DbSession session) throws SQLException {
+    PreparedStatement stmt = session.getConnection().prepareStatement(SQL_METRICS);
+    stmt.setBoolean(1, true);
+    return stmt;
+  }
+
+  private static List<Project> selectProjects(DbSession session, long afterDate, @Nullable String projectUuid) {
+    List<Project> projects = new ArrayList<>();
+    try (PreparedStatement stmt = createProjectsStatement(session, afterDate, projectUuid);
+      ResultSet rs = stmt.executeQuery()) {
+      while (rs.next()) {
+        String orgUuid = rs.getString(1);
+        String uuid = rs.getString(2);
+        String key = rs.getString(3);
+        String name = rs.getString(4);
+        String analysisUuid = DatabaseUtils.getString(rs, 5);
+        Long analysisDate = DatabaseUtils.getLong(rs, 6);
+        Project project = new Project(orgUuid, uuid, key, name, analysisUuid, analysisDate);
+        projects.add(project);
+      }
+      return projects;
+    } catch (SQLException e) {
+      throw new IllegalStateException("Fail to execute request to select all projects", e);
+    }
+  }
+
+  private static PreparedStatement createProjectsStatement(DbSession session, long afterDate, @Nullable String projectUuid) {
+    try {
+      String sql = SQL_PROJECTS;
+      sql += afterDate <= 0L ? "" : DATE_FILTER;
+      sql += projectUuid == null ? "" : PROJECT_FILTER;
+      PreparedStatement stmt = session.getConnection().prepareStatement(sql);
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png 'Severity: BLOCKER') "" is provided externally to the method and not sanitized before use. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2077)